File: SRM_747_DivII_Problem_1.py
# code_report Solution
# Video Link: https://youtu.be/g0MsAyI_zGI
# Problem Link: https://arena.topcoder.com/#/u/practiceCode/17404/78412/15285/2/332030

class Sherlock:
    def isItHim(self, f, l):
        isHim = len(f) >= 7 and \
                len(l) >= 7 and \
                f[0] == 'B' and \
                l[0] == 'C' and \
                sum(1 for i in f if i in 'BENEDICT')   >= 3 and \
                sum(1 for i in l if i in 'CUMERBATCH') >= 5
        if isHim: return 'It is him'
        else:     return 'It is someone else'

File: SRM_747_DivII_Problem_1.cpp
// code_report Solution
// Video Link: https://youtu.be/g0MsAyI_zGI
// Problem Link: https://arena.topcoder.com/#/u/practiceCode/17404/78412/15285/2/332030

#include <vector>
#include <numeric>
#include <string>
#include <algorithm>
#include <set>

using namespace std;

class Sherlock {
public:
    string isItHim(string f, string l) {
        set<char> b = {'B','E','N','D','I','C','T'};
        set<char> c = {'C','U','M','B','E','R','A','T','H'};
        auto isHim = f.size() >= 7 && 
                     l.size() >= 7 && 
                     f[0] == 'B'   &&
                     l[0] == 'C'   && 
                     count_if(f.begin(), f.end(), [&b](char e) { return b.count(e); }) >= 3 && 
                     count_if(l.begin(), l.end(), [&c](char e) { return c.count(e); }) >= 5;
        return isHim ? "It is him" : "It is someone else";
    }
};
